utils.py

```diff
@@ -5,17 +5,247 @@
 """
 
 import logging
+import os
 import os.path as op
 import re
+import shutil
+from ast import literal_eval as leval
 from collections import OrderedDict
+from pathlib import Path
+from zipfile import ZipFile
 
+import dicom2nifti
 import nibabel as nib
 import numpy as np
+import pydicom
+import requests
 from skimage import draw
 
 log = logging.getLogger(__name__)
 
 
+class InvalidDICOMFile(Exception):
+    """Exception raised when an invalid DICOM file is encountered.
+
+    Attributes:
+        expression -- input expression in which the error occurred
+        message -- explanation of the error
+    """
+
+    def __init__(self, message):
+        Exception.__init__(self)
+        self.message = message
+
+
+class InvalidROIError(Exception):
+    """Exception raised when session ROI data is invalid.
+
+    Attributes:
+        expression -- input expression in which the error occurred
+        message -- explanation of the error
+    """
+
+    def __init__(self, message):
+        Exception.__init__(self)
+        self.message = message
+
+
+def io_proxy_wado(
+    api_key, api_key_prefix, project_id, study=None, series=None, instance=None
+):
+    """
+    Request wrapper for io-proxy api (https://{instance}/io-proxy/docs#/)
+
+    Args:
+        api_key (str): Full instance api-key
+        api_key_prefix (str): Type of user (e.g. 'scitran-user')
+        project_id (str): Project ID to inquire
+        study (str, optional): DICOM StudyUID. Defaults to None.
+        series (str, optional): DICOM SeriesUID. Defaults to None.
+        instance (str, optional): DICOM InstanceUID. Defaults to None.
+
+    Returns:
+        dict/list: A dictionary for dicom tags or a list of dictionaries with dicom tags
+    """
+    base_url = Path(api_key.split(":")[0])
+    base_url /= "io-proxy/wado"
+    base_url /= "projects/" + project_id
+    if study:
+        base_url /= "studies/" + study
+    if series:
+        base_url /= "series/" + series
+        base_url /= "instances"
+    if instance:
+        base_url /= instance
+        base_url /= "tags"
+    base_url = "https://" + str(base_url)
+
+    headers = {
+        "Authorization": api_key_prefix + " " + api_key,
+        "accept": "application/json",
+    }
+
+    req = requests.get(base_url, headers=headers)
+
+    return leval(req.text)
+
+
+def convert_ROI_to_nifti_form(
+    fw_client, project_id, file_obj, imagePath, ohifViewer_info
+):
+    """
+    Converts ohifViewer ROI from a dicom to that of an individual nifti format.
+
+    Args:
+        fw_client (flywheel.Client): The active flywheel client
+        project_id (str): The id of the project to search for io-proxy elements
+        file_obj (flywheel.File): File object data
+        imagePath (str): The DICOM representation of the ohifViewer ROI imagePath
+        ohifViewer_info (dict): The full session-level ohifViewer ROI dictionary.
```

Comment:
Maybe explain why this needs to be passed in here.  I expect that it has higher level ROI info and then `convert_ROI_to_nifti_form()` fills in the actual outlines?  This is probably my only chance to ask big-picture questions.

Reply:
The current ohifViewer annotations on dicoms end up in the session info.

Reply:
Questions after next week will be $$$$.